upgrades/steps_27.go

```diff
@@ -54,6 +58,28 @@ func stateStepsFor27() []Step {
 			run: func(context Context) error {
 				return context.State().EnsureRelationApplicationSettings()
 			},
+		}, &upgradeStep{
+			description: "change owner of unit and machine logs to adm",
+			targets:     []Target{AllMachines},
+			run: func(context Context) error {
+				return context.State().AddModelLogfileMaxSize()
+			},
+		}, &upgradeStep{
+			description: "change owner of unit and machine logs to adm",
+			targets:     []Target{AllMachines},
+			run:         resetLogPermissions,
 		},
 	}
 }
+
+// This adds upgrade steps, we just rewrite the default values which are set before.
+// With this we can make sure that things are changed in one default place
+func resetLogPermissions(context Context) error {
+	sysdManager := service.NewServiceManagerWithDefaults()
```

Comment:
Does this actually change the ownership of the existing log files?

Reply:
My upgrade testing shows that it doesn't write the files for the agents on upgrade.

Reply:
So what it does it to overwrite the existing systemd configuration files. 
```
ubuntu@juju-79b12a-0:/lib/systemd/system$ cd jujud-machine-0/
ubuntu@juju-79b12a-0:/lib/systemd/system/jujud-machine-0$ cat exec-start.sh 
#!/usr/bin/env bash

# Set up logging.
touch '/var/log/juju/machine-0.log'
chown syslog:adm '/var/log/juju/machine-0.log'
chmod 0640 '/var/log/juju/machine-0.log'
exec >> '/var/log/juju/machine-0.log'
exec 2>&1

# Run the script.
'/var/lib/juju/tools/machine-0/jujud' machine --data-dir '/var/lib/juju' --machine-id 0 --debug

ubuntu@juju-79b12a-0:/lib/systemd/system$ cd jujud-unit-ubuntu-0/
ubuntu@juju-79b12a-0:/lib/systemd/system/jujud-unit-ubuntu-0$ ls
exec-start.sh  jujud-unit-ubuntu-0.service
ubuntu@juju-79b12a-0:/lib/systemd/system/jujud-unit-ubuntu-0$ cat exec-start.sh 
#!/usr/bin/env bash

# Set up logging.
touch '/var/log/juju/unit-ubuntu-0.log'
chown syslog:adm '/var/log/juju/unit-ubuntu-0.log'
chmod 0640 '/var/log/juju/unit-ubuntu-0.log'
exec >> '/var/log/juju/unit-ubuntu-0.log'
exec 2>&1

# Run the script.
'/var/lib/juju/tools/unit-ubuntu-0/jujud' unit --data-dir '/var/lib/juju' --unit-name ubuntu/0 --debug
```

What I should add is a kind of agent restart.

Or did you meant something else as well?

Reply:
@howbazaar so was the change now should make sure that everyone got the correct systemd configuration files. Though I did intentionally not add the restarts because I was not sure how this affects the whole system. What do you think would be the best action here?

Reply:
There is a slight problem here in that we will try to use the systemd service manager even if the machine/unit is on trusty (upstart) or windows.

To simplify situations, we need some extra checks in the method.

We need to skip this upgrade step if one of the following is true:
 * this is a caas controller agent
 * this is a unit agent (the machine rewrites all the config files)
 * if the OS is not linux
 * if the discovered service mangager isn't systemd

Probably easiest to do a service manager discovery at the top of the function and skip for non-systemd. This would catch both old ubuntu and windows in one go.

You can check the Tag() type from the agent config to determine whether it is a machine agent or unit agent. Skip for unit agents because the machine agent is doing all the work.

I'm not entirely sure how to identify if the agent is a caas controller or not. That would take some looking into to check the agent.conf for the different types to work out what is different.

Reply:
:+1:

Reply:
:+1: , I did not take a look into our caas controller. Why do we want to skip those? I mostly looked into steps_24. Caas was not skipped there.
Do we have a special setup there, which we do not want to be changed?

Additionally, did you mean changing permissions of the whole (/var/log/juju) dir, as I was initially planning only to change unit and machine logs.